File: clients/r/R/a2a.R
# ---------------------------------------------------------------------------
# A2A (agent-to-agent) JSON-RPC + agent discovery
# ---------------------------------------------------------------------------
#
# The node exposes an A2A JSON-RPC 2.0 endpoint at POST /api/a2a with methods
# message/send, tasks/get and tasks/list, an ES256-signed agent card at
# /.well-known/agent-card.json and the verifying key set at
# /.well-known/jwks.json.

#' Fetch the signed agent card
#'
#' Calls `GET /.well-known/agent-card.json` and returns the agent card as a
#' parsed list. The card advertises the agent's identity, A2A endpoint and
#' skills and is signed with ES256 (the verifying key is published at
#' `/.well-known/jwks.json`). This client returns the card as served; it does
#' not verify the JWS signature.
#'
#' @param base_url OABP base URL. Defaults to [oabp_base_url()].
#'
#' @return A named list: the parsed agent card.
#' @export
#' @examples
#' \dontrun{
#' card <- oabp_agent_card()
#' card$name
#' card$url        # the A2A JSON-RPC endpoint
#' }
oabp_agent_card <- function(base_url = oabp_base_url()) {
  req <- .oabp_req(".well-known/agent-card.json", base_url = base_url)
  .oabp_perform(req)
}

#' Low-level A2A JSON-RPC call
#'
#' Performs a single JSON-RPC 2.0 request against `POST /api/a2a`. Most callers
#' will prefer the typed wrappers [oabp_a2a_message_send()],
#' [oabp_a2a_task_get()] and [oabp_a2a_tasks_list()].
#'
#' @param method JSON-RPC method name, e.g. `"message/send"`, `"tasks/get"`,
#'   `"tasks/list"`.
#' @param params Named list of method parameters (sent as the JSON-RPC
#'   `params` member). Defaults to an empty object.
#' @param id JSON-RPC request id. Defaults to a fresh value.
#' @param token Optional bearer token if the endpoint requires authentication.
#' @param base_url OABP base URL. Defaults to [oabp_base_url()].
#'
#' @return The JSON-RPC `result` member (parsed) on success. On a JSON-RPC
#'   `error` member, an R error is raised carrying the error `code` and
#'   `message`.
#' @export
#' @examples
#' \dontrun{
#' oabp_a2a_request("tasks/list")
#' }
oabp_a2a_request <- function(method, params = list(), id = NULL,
                             token = NULL, base_url = oabp_base_url()) {
  .check_string(method, "method")
  if (!is.list(params)) {
    stop("`params` must be a (possibly empty) named list.", call. = FALSE)
  }
  if (is.null(id)) {
    id <- as.character(as.integer(Sys.time()))
  }

  payload <- list(
    jsonrpc = "2.0",
    id      = id,
    method  = method,
    params  = .as_json_object(params)
  )

  req <- .oabp_req("api/a2a", base_url = base_url, token = token)
  req <- httr2::req_method(req, "POST")
  req <- httr2::req_body_json(req, payload, auto_unbox = TRUE)
  body <- .oabp_perform(req)

  if (!is.null(body[["error"]])) {
    err <- body[["error"]]
    code <- .scalar(err[["code"]], NA)
    msg  <- .scalar(err[["message"]], "unknown JSON-RPC error")
    stop(sprintf("A2A JSON-RPC error %s: %s", code, msg), call. = FALSE)
  }
  body[["result"]] %||% list()
}

#' Send a message to an agent over A2A
#'
#' Wraps the `message/send` JSON-RPC method. Builds a minimal A2A message
#' (one text part) and sends it to the node, returning the resulting task /
#' message object.
#'
#' @param text The message text.
#' @param role Message role, defaults to `"user"`.
#' @param task_id Optional existing task id to continue a conversation.
#' @param message_id Optional client-supplied message id. Defaults to a fresh
#'   value.
#' @param token Optional bearer token.
#' @param base_url OABP base URL. Defaults to [oabp_base_url()].
#'
#' @return The parsed JSON-RPC result (typically a task object).
#' @export
#' @examples
#' \dontrun{
#' task <- oabp_a2a_message_send("List the open USDC missions, please.")
#' task$id
#' }
oabp_a2a_message_send <- function(text, role = "user", task_id = NULL,
                                  message_id = NULL, token = NULL,
                                  base_url = oabp_base_url()) {
  .check_string(text, "text")
  if (is.null(message_id)) {
    message_id <- paste0("msg-", as.integer(Sys.time()))
  }
  message <- list(
    role      = role,
    parts     = list(list(kind = "text", text = text)),
    messageId = message_id,
    kind      = "message"
  )
  if (!is.null(task_id)) {
    message[["taskId"]] <- task_id
  }
  oabp_a2a_request("message/send", params = list(message = message),
                   token = token, base_url = base_url)
}

#' Fetch an A2A task by id
#'
#' Wraps the `tasks/get` JSON-RPC method.
#'
#' @param task_id Task id to fetch.
#' @param token Optional bearer token.
#' @param base_url OABP base URL. Defaults to [oabp_base_url()].
#'
#' @return The parsed task object.
#' @export
#' @examples
#' \dontrun{
#' oabp_a2a_task_get("task_123")
#' }
oabp_a2a_task_get <- function(task_id, token = NULL, base_url = oabp_base_url()) {
  .check_string(task_id, "task_id")
  oabp_a2a_request("tasks/get", params = list(id = task_id),
                   token = token, base_url = base_url)
}

#' List A2A tasks
#'
#' Wraps the `tasks/list` JSON-RPC method.
#'
#' @param token Optional bearer token.
#' @param base_url OABP base URL. Defaults to [oabp_base_url()].
#'
#' @return The parsed result of `tasks/list` (typically a list of task
#'   objects). Returned as-is from the node.
#' @export
#' @examples
#' \dontrun{
#' oabp_a2a_tasks_list()
#' }
oabp_a2a_tasks_list <- function(token = NULL, base_url = oabp_base_url()) {
  oabp_a2a_request("tasks/list", token = token, base_url = base_url)
}

File: clients/go/a2a.go
package oabp

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync/atomic"
)

// A2A (Agent-to-Agent) support: a JSON-RPC 2.0 client over POST /api/a2a, plus
// helpers to fetch the ES256-signed agent card and the JWKS used to verify it.

// pathA2A is the JSON-RPC endpoint.
const pathA2A = "/api/a2a"

// JSON-RPC method names exposed by the protocol's A2A endpoint.
const (
	A2AMethodMessageSend = "message/send"
	A2AMethodTasksGet    = "tasks/get"
	A2AMethodTasksList   = "tasks/list"
)

// rpcRequest is a JSON-RPC 2.0 request envelope.
type rpcRequest struct {
	JSONRPC string `json:"jsonrpc"`
	ID      int64  `json:"id"`
	Method  string `json:"method"`
	Params  any    `json:"params,omitempty"`
}

// rpcError is a JSON-RPC 2.0 error object.
type rpcError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// Error implements error.
func (e *rpcError) Error() string {
	if len(e.Data) > 0 {
		return fmt.Sprintf("oabp a2a: rpc error %d: %s: %s", e.Code, e.Message, string(e.Data))
	}
	return fmt.Sprintf("oabp a2a: rpc error %d: %s", e.Code, e.Message)
}

// rpcResponse is a JSON-RPC 2.0 response envelope. Result is left raw so callers
// can decode it into the appropriate type.
type rpcResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Result  json.RawMessage `json:"result"`
	Error   *rpcError       `json:"error"`
}

// rpcID is a monotonically increasing JSON-RPC id source. A single atomic
// counter is sufficient because ids only need to be unique per client.
var rpcID atomic.Int64

// A2ACall performs a raw JSON-RPC call against the A2A endpoint and decodes the
// result into out (which may be nil to discard it). It returns the JSON-RPC
// error as an *rpcError when the server reports one.
//
// Most callers should use the typed helpers (SendMessage, GetTask, ListTasks)
// instead; A2ACall is exported for forward-compatibility with methods this SDK
// version does not yet model.
func (c *Client) A2ACall(ctx context.Context, method string, params, out any) error {
	reqEnv := rpcRequest{
		JSONRPC: "2.0",
		ID:      rpcID.Add(1),
		Method:  method,
		Params:  params,
	}
	var respEnv rpcResponse
	if err := c.doJSON(ctx, http.MethodPost, pathA2A, reqEnv, &respEnv); err != nil {
		return err
	}
	if respEnv.Error != nil {
		return respEnv.Error
	}
	if out == nil || len(respEnv.Result) == 0 {
		return nil
	}
	if err := json.Unmarshal(respEnv.Result, out); err != nil {
		return fmt.Errorf("oabp a2a: decode result of %q: %w", method, err)
	}
	return nil
}

// A2AMessage is an A2A message, following the A2A protocol shape: a role and an
// ordered list of parts. Use TextMessage for the common single-text case.
type A2AMessage struct {
	Role      string    `json:"role"`
	Parts     []A2APart `json:"parts"`
	MessageID string    `json:"messageId,omitempty"`
	Metadata  any       `json:"metadata,omitempty"`
}

// A2APart is one part of a message. Kind is "text" for textual parts; the Text
// field carries the content. Other kinds (file/data) pass through Raw.
type A2APart struct {
	Kind string `json:"kind"`
	Text string `json:"text,omitempty"`
}

// TextMessage builds a user-role A2A message carrying a single text part.
func TextMessage(text string) A2AMessage {
	return A2AMessage{
		Role:  "user",
		Parts: []A2APart{{Kind: "text", Text: text}},
	}
}

// A2ATask is a task returned by the A2A endpoint. The protocol's task objects
// vary in their fields across implementations, so the documented core (id /
// status) is typed and the full object is retained in Raw.
type A2ATask struct {
	ID     string          `json:"id"`
	Status json.RawMessage `json:"status,omitempty"`
	Raw    json.RawMessage `json:"-"`
}

// UnmarshalJSON decodes an A2ATask and keeps the raw object in Raw.
func (t *A2ATask) UnmarshalJSON(data []byte) error {
	type alias A2ATask
	var a alias
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	*t = A2ATask(a)
	raw := make(json.RawMessage, len(data))
	copy(raw, data)
	t.Raw = raw
	return nil
}

// SendMessage invokes the A2A "message/send" method and returns the raw result,
// which per the A2A spec is either a Message or a Task depending on the agent.
// The bytes are returned verbatim so callers can decode whichever they expect.
func (c *Client) SendMessage(ctx context.Context, msg A2AMessage) (json.RawMessage, error) {
	params := map[string]any{"message": msg}
	var result json.RawMessage
	if err := c.A2ACall(ctx, A2AMethodMessageSend, params, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// GetTask invokes the A2A "tasks/get" method for a given task id.
func (c *Client) GetTask(ctx context.Context, taskID string) (*A2ATask, error) {
	if taskID == "" {
		return nil, fmt.Errorf("oabp a2a: GetTask: empty task id")
	}
	params := map[string]any{"id": taskID}
	var task A2ATask
	if err := c.A2ACall(ctx, A2AMethodTasksGet, params, &task); err != nil {
		return nil, err
	}
	return &task, nil
}

// ListTasks invokes the A2A "tasks/list" method and returns the tasks. Some
// deployments wrap the array in a {"tasks":[...]} object; both shapes are
// handled.
func (c *Client) ListTasks(ctx context.Context) ([]A2ATask, error) {
	var raw json.RawMessage
	if err := c.A2ACall(ctx, A2AMethodTasksList, map[string]any{}, &raw); err != nil {
		return nil, err
	}
	// Try a bare array first.
	var tasks []A2ATask
	if err := json.Unmarshal(raw, &tasks); err == nil {
		return tasks, nil
	}
	// Fall back to a wrapped object.
	var wrapped struct {
		Tasks []A2ATask `json:"tasks"`
	}
	if err := json.Unmarshal(raw, &wrapped); err != nil {
		return nil, fmt.Errorf("oabp a2a: decode tasks/list result: %w", err)
	}
	return wrapped.Tasks, nil
}

// AgentCard is the (partial) A2A agent card describing this protocol's agent.
// Only the commonly used fields are typed; the complete signed document is in
// Raw for callers that verify the ES256 signature against the JWKS.
type AgentCard struct {
	Name               string          `json:"name"`
	Description        string          `json:"description,omitempty"`
	URL                string          `json:"url,omitempty"`
	Version            string          `json:"version,omitempty"`
	ProtocolVersion    string          `json:"protocolVersion,omitempty"`
	PreferredTransport string          `json:"preferredTransport,omitempty"`
	Raw                json.RawMessage `json:"-"`
}

// UnmarshalJSON decodes an AgentCard and keeps the raw document in Raw so the
// ES256 JWS signature can be verified by the caller.
func (a *AgentCard) UnmarshalJSON(data []byte) error {
	type alias AgentCard
	var v alias
	if err := json.Unmarshal(data, &v); err != nil {
		return err
	}
	*a = AgentCard(v)
	raw := make(json.RawMessage, len(data))
	copy(raw, data)
	a.Raw = raw
	return nil
}

// AgentCard fetches the ES256-signed agent card from
// /.well-known/agent-card.json.
func (c *Client) AgentCard(ctx context.Context) (*AgentCard, error) {
	var card AgentCard
	if err := c.doJSON(ctx, http.MethodGet, "/.well-known/agent-card.json", nil, &card); err != nil {
		return nil, err
	}
	return &card, nil
}

// JWKS fetches the JSON Web Key Set used to verify the agent card's ES256
// signature, from /.well-known/jwks.json. The keys are returned as raw JSON
// objects so callers can feed them to their JWK library of choice.
func (c *Client) JWKS(ctx context.Context) ([]json.RawMessage, error) {
	var set struct {
		Keys []json.RawMessage `json:"keys"`
	}
	if err := c.doJSON(ctx, http.MethodGet, "/.well-known/jwks.json", nil, &set); err != nil {
		return nil, err
	}
	return set.Keys, nil
}
